### PR TITLE
Boost rtree wrapper

### PR DESCRIPTION
**Pull request by luca-heltai**

### CODE REVIEW

File: include/deal.II/numerics/rtree.h [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+#include <memory>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * A wrapper for the boost::geometry library, used to compute an RTree from a
+ * collection of Point<dim> based spatial objects, BoundingBox<dim>, or
+ * Segment<dim> base spatial objects.
```
Comment (jppelteret) [inline include/deal.II/numerics/rtree.h]: Might it be worth explaining briefly what an RTree is, and what it might be useful for?

File: include/deal.II/numerics/rtree.h [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+ *
+ * An RTree may store any type of @p LeafType as long as it is possible to extract
+ * an @p Indexable that the RTree can handle and compare Values. The @p Indexable
+ * is a type adapted to Point, BoundingBox or Segment concept.
+ *
+ * In particular, given an Indexable (Point,  BoundingBox or Segment), @p Leaftype
```
Comment (jppelteret) [inline include/deal.II/numerics/rtree.h]: `@p Leaftype` -> `@p LeafType`

Comment (jppelteret) [reply include/deal.II/numerics/rtree.h]: > given an Indexable (Point,  BoundingBox or Segment)

an Indexable -> `an @p Indexable`.

Also, this clause seems to be incomplete. Do you mean to say something like `given an @p Indexable type (for example, a Point, BoundingBox or Segment)`?

File: include/deal.II/numerics/rtree.h [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+ * can by any of @p Indexable,  std::pair<Indexable, T>,
+ * boost::tuple<Indexable, ...> or std::tuple<Indexable, ...>.
+ *
+ * The optional argument @p IndexType is used only when adding elements to the
+ * tree one by one. If a range insertion is used, then the tree is built using
+ * the packing algorithm. Linear, quadratic, and rstar algorithms are available
+ * if one wants to construct the tree sequentially, however none of these is
+ * very efficient, and users should use the packing algorithm when possible,
+ * which construct the tree all at once, when you have all the leafs at your
+ * disposal. In this case, the @p IndexType parameter is not used, and you can
```
Comment (jppelteret) [inline include/deal.II/numerics/rtree.h]: This is a long sentence that's a bit hard to follow. May I suggest this minor reformulation (note the spelling corrections too):

Linear, quadratic, and rstar algorithms are available if one wants to construct the tree sequentially. 
However, none of these is very efficient, and users should use the packing algorithm when possible.
The packing algorithm constructs the tree all at once, and may be used when you have all the leaves at your disposal.

File: include/deal.II/numerics/rtree.h [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+ * safely use the default.
+ *
+ * This class is usually used in combination with one of the two helper
+ * functions pack_rtree(), that take a container or a range of iterators to
```
Comment (jppelteret) [inline include/deal.II/numerics/rtree.h]: that takes a

Comment (jppelteret) [reply include/deal.II/numerics/rtree.h]: iterators to construct -> iterators and constructs

File: include/deal.II/numerics/rtree.h [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+ * @code
+ * std::vector<Point<2>> points = generate_some_points();
+ * auto tree = pack_rtree(points.begin(), points.end());
+ * // or, equivalently:
+ * // auto tree = pack_rtree(points);
+ * @endcode
+ *
+ * The tree is accessed by using boost::geometry::index queries (see
+ * https://www.boost.org/doc/libs/1_68_0/libs/geometry/doc/html/geometry/spatial_indexes/queries.html
+ * for some examples and extensive documentations on the queries that you can
```
Comment (jppelteret) [inline include/deal.II/numerics/rtree.h]: documentations -> documentation

File: include/deal.II/numerics/rtree.h [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * A wrapper for the boost::geometry library, used to compute an RTree from a
+ * collection of Point<dim> based spatial objects, BoundingBox<dim>, or
+ * Segment<dim> base spatial objects.
+ *
+ * An RTree may store any type of @p LeafType as long as it is possible to extract
+ * an @p Indexable that the RTree can handle and compare Values. The @p Indexable
+ * is a type adapted to Point, BoundingBox or Segment concept.
```
Comment (jppelteret) [inline include/deal.II/numerics/rtree.h]: to the Point

Comment (jppelteret) [reply include/deal.II/numerics/rtree.h]: What makes something Indexable? Is it that it has the `<` comparator defined? If so, maybe its worthwhile mentioning this.

File: include/deal.II/numerics/rtree.h [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * A wrapper for the boost::geometry library, used to compute an RTree from a
+ * collection of Point<dim> based spatial objects, BoundingBox<dim>, or
+ * Segment<dim> base spatial objects.
+ *
+ * An RTree may store any type of @p LeafType as long as it is possible to extract
+ * an @p Indexable that the RTree can handle and compare Values. The @p Indexable
```
Comment (jppelteret) [inline include/deal.II/numerics/rtree.h]: `The @p Indexable` -> `An @p Indexable`

Comment (jppelteret) [reply include/deal.II/numerics/rtree.h]: I've not found it particularly easy to follow the use of the term `Indexable` throughout this documentation. Might it make more sense to call these `IndexableType`s?

File: include/deal.II/numerics/rtree.h [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+ */
+template <typename LeafType,
+          typename IndexType = boost::geometry::index::linear<16>>
+using RTree = boost::geometry::index::rtree<LeafType, IndexType>;
+
+/**
+ * Construct the correct RTree object by passing an iterator range.
+ *
+ * Notice that the order of the parameters is the opposite with respect to the
+ * RTree class, since we can automatically infer the LeafType from the
```
Comment (jppelteret) [inline include/deal.II/numerics/rtree.h]: `@p LeafType`

File: include/deal.II/numerics/rtree.h [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+ * arguments, and we only need to specify the @p IndexType if the default is not
+ * already good.
```
Comment (jppelteret) [inline include/deal.II/numerics/rtree.h]: > is not already good

is not adequate (?) 

Same applies below.

File: include/deal.II/numerics/rtree.h [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+RTree<typename ContainerType::value_type, IndexType>
+pack_rtree(const ContainerType &container);
+
+
+
+// Inline and template functions
+#ifndef DOXYGEN
+template <typename IndexType, typename LeafTypeIterator>
+RTree<typename LeafTypeIterator::value_type, IndexType>
+pack_rtree(LeafTypeIterator begin, LeafTypeIterator end)
```
Comment (jppelteret) [inline include/deal.II/numerics/rtree.h]: Can we mark these input iterators as `const`?

File: tests/boost/rtree_02.cc [hunk 11]
```diff
@@ -0,0 +1,65 @@
+/* ---------------------------------------------------------------------
+ *
+ * Copyright (C) 1999 - 2018 by the deal.II authors
```
Comment (jppelteret) [inline tests/boost/rtree_02.cc]: Gosh, you've had this test lying around for a long time ;-P

File: doc/news/changes/minor/20181006LucaHeltai [hunk 12]
```diff
@@ -0,0 +1,5 @@
+New: Added RTree class and pack_rtree() function to create boost::gemoetry::index::rtree from
```
Comment (jppelteret) [inline doc/news/changes/minor/20181006LucaHeltai]: to create a

File: include/deal.II/numerics/rtree.h (hunk 7 above)
Comment (luca-heltai) [reply include/deal.II/numerics/rtree.h]: I'm not sure. Boost uses Indexable in all its documentation. Changing name here would only make it more confusing... I tried to extend a little bit the explanation at the beginning. Maybe this will make you happier. :)

File: include/deal.II/numerics/rtree.h (hunk 5 above)
Comment (luca-heltai) [reply include/deal.II/numerics/rtree.h]: it's fixed, but does not show...

File: include/deal.II/numerics/rtree.h (hunk 7 above)
Comment (jppelteret) [reply include/deal.II/numerics/rtree.h]: > Boost uses Indexable in all its documentation. Changing name here would only make it more confusing...

Fair enough. Consistency between our documentation and there's would be a good thing.

> I tried to extend a little bit the explanation at the beginning. Maybe this will make you happier. :)

Yes, the additional documentation is really helpful. Thanks!

Comment (jppelteret) [reply include/deal.II/numerics/rtree.h]: Still, it remains that we need to adjust one of the articles used in this sentence. I don't think that its correct to use both a definite and an indefinite article when referencing the same thing. So this should either read

> The `@p Indexable` is the type adapted to the Point, BoundingBox or Segment concept

or 

> An `@p Indexable` is a type adapted to the Point, BoundingBox or Segment concept

I favour the second option, since it refers to the abstract concept in the same way that you have done in the preceding sentence.

File: include/deal.II/numerics/rtree.h [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * A wrapper for the boost::geometry::index::rtree class, implementing a
+ * self-balancing spatial index (the R-tree) capable of storing various types of
+ * values, using different balancing algorithms.
+ *
+ * From wikipedia:
```
Comment (jppelteret) [inline include/deal.II/numerics/rtree.h]: wikipedia -> Wikipedia

Comment (jppelteret) [reply include/deal.II/numerics/rtree.h]: Thank you for adding this quotation. I think its really helpful to have this explanation of what an R-tree is.

May we please add a hyperlink to the Wikipedia article?

File: include/deal.II/numerics/rtree.h [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+ * insert into the subtree that requires least enlargement of its bounding box.
+ * Once that page is full, the data is split into two sets that should cover the
+ * minimal area each. Most of the research and improvements for R-trees aims at
+ * improving the way the tree is built and can be grouped into two objectives:
+ * building an efficient tree from scratch (known as bulk-loading) and
+ * performing changes on an existing tree (insertion and deletion).
+ * @endquotation
+ *
+ * An RTree may store any type of @p LeafType as long as it is possible to extract
+ * an @p Indexable that the RTree can handle and compare Values. The @p Indexable
```
Comment (jppelteret) [inline include/deal.II/numerics/rtree.h]: compare Values -> compare values

File: tests/boost/rtree_02.cc [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * The deal.II library is free software; you can use it, redistribute
+ * it, and/or modify it under the terms of the GNU Lesser General
+ * Public License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ * The full text of the license can be found in the file LICENSE.md at
+ * the top level directory of deal.II.
+ *
+ * ---------------------------------------------------------------------
+
```
Comment (jppelteret) [inline tests/boost/rtree_02.cc]: Stray new line

File: tests/boost/rtree_02.cc [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+ */
+
```
Comment (jppelteret) [inline tests/boost/rtree_02.cc]: Hmm... There's no explanation as to what the test is used for. Sorry I missed that on the first review. Would you mind please adding one?

File: include/deal.II/numerics/rtree.h [hunk 17]
```diff
@@ -36,7 +36,7 @@ DEAL_II_NAMESPACE_OPEN
  * self-balancing spatial index (the R-tree) capable of storing various types of
  * values, using different balancing algorithms.
  *
- * From wikipedia:
+ * From Wikipedia (https://en.wikipedia.org/wiki/R-tree):
```
Comment (jppelteret) [inline include/deal.II/numerics/rtree.h]: We typically add links as hyperlinks (sorry that I wasn't explicit here):
```
From <a href="https://en.wikipedia.org/wiki/R-tree">Wikipedia</a>:
```

Comment (luca-heltai) [reply include/deal.II/numerics/rtree.h]: There is no need. Doxygen supports markdown, which translates links to links explicitly, unless you want the more aesthetically pleasing link [Wikipedia](https://en.wikipedia.org/wiki/R-tree)

**Comment by jppelteret** [comment]
@luca-heltai Thanks for fixing those links. I guess that if nobody has any comments by tomorrow then this can be squashed and merged.
